enhancements/installer/kubernete-slb-owned-api-loadbalancers.md

```diff
@@ -0,0 +1,508 @@
+
+The installer terminates the `bootstrap` machine and also removes the `bootstrap` machine from the backend of all the loadbalancers. And only 
+the `master-{0,1,2}` machines remain.
+
+During all these stages, all the clients end up using the loadbalancers as a single abstraction and do not have to understand the bootstrapping 
+stages or even notice the bootstrap-control-plane.
+
+## Proposal
+
+During bootstrapping the clients use the known API, and API-INT DNS names to discover the endpoints. Ensuring that the clients continue to depend on 
+that abstraction is very important as it allows the clients to be simple.
+
+Secondly, the load balancers cannot be created by Kubernetes cloud controller manager unless there is a Kubernetes API already running, but for 
+successful bootstrapping we need the loadbalancers so that the kubelets can communicate with bootstrap-control-plane to run the necessary pods to create 
+the k8s API.
+
+So, the break the cyclic dependency, I suggest we modify the stages as follows,
+
+**Stage 1:** Bootstrapping control plane on bootstrap host
+
+PUBLIC ZONE:api=static-public-lb-ip
+PRIVATE ZONE:api=bootstrap-private-ip,static-private-lb-ip;api-int=bootstrap-private-ip,static-private-lb-ip
+
+The `bootstrap` machine hosts the bootstrap-control-plane, and the installer has pre-seeded the DNS entries with reserved ips for loadbalancers to 
+be created by the cluster. The kubelet on `master-{0,1,2}` use the DNS to resolve the endpoints to the `bootstrap` machine, 
+and communicate with the k8s API.
+
+An entity generates Kubernetes service type loadbalancer objects using the static ips created by the installer and pushes them to the cluster using 
+the bootstrap control plane. The bootstrap-control-plane 's kube-controller-manager creates the LBs for the services.
+
+**Stage 2:** Bootstrapping complete
+
+PUBLIC ZONE:api=static-public-lb-ip
+PRIVATE ZONE:api=bootstrap-private-ip,static-private-lb-ip;api-int=bootstrap-private-ip,static-private-lb-ip
+
+public-lb:master-0(green);master-1(green);master-2(green)
+private-lb:master-0(green);master-1(green);master-2(green)
+
+Bootstrapping of the k8s API is complete when the necessary pods are running, and the service type load balancers are running. At this point 
+the `bootstrap-control-plane` shuts itself down. The clients can now reach the k8s API using the loadbalancer ips, the clients trying to reach API 
+using `bootstrap` host will fail and round-robin to the loadbalancer ips.
+
+**Stage 3:** Bootstrap host removed
+
+PUBLIC ZONE:api=static-public-lb-ip
+PRIVATE ZONE:api=static-private-lb-ip;api-int=static-private-lb-ip
+
+public-lb:master-0(green);master-1(green);master-2(green)
+private-lb:master-0(green);master-1(green);master-2(green)
+
+The installer terminates the `bootstrap` machine and also removes the `bootstrap` machine's ip addresses from the DNS. And only the `master-{0,1,2}` machines
+remain with only loadbalancer ips in the DNS.
+
+### User Stories
```

Comment:
I'd like to see a user story to explicitly support private clusters:

> As an OpenShift Dedicated customer, I would like (continue to be able) to disallow all external traffic to my cluster's API server, while continuing to allow traffic from inside the private cloud network.

This has two, twin paired user stories which must be considered at the same time:

> As a member of a Red Hat SRE team, I need to have access to customers' private clusters so that I can support the cluster.
> As a Hive process, I need to have access to customers' private clusters so that I can manage the cluster

For background, we (SREP) create a second `Service` of type `LoadBalancer` for the API server (`rh-api`) which limits access to Red Hat controlled IPs. Hive "latches" on to this endpoint once it becomes available and uses it for _all_ cluster management (public and private). We delete the external `api.<cluster-domain>` NLB and adjust DNS for set a cluster "private"

Presently, these are handled by a combination of [cloud-ingress-operator](https://github.com/openshift/cloud-ingress-operator), OCM, and Hive (apiurloverride, if I recall the attribute correctly). It sounds very much like we could retire some of these pieces if we could support these user stories within this enhancement request. How would that look?

If there was first class support to create an additional "external" `LoadBalancer` `Service` at install time for the API, we (SREP) could populate its `loadBalancerSourceRanges` with the list of Red Hat controlled IPs at a later time (eg post-install day 2) via a Hive syncset. When a cluster is set private ("internal"), deleting the external `api.<cluster-domain>` `Service` would still effectively let us "seal off" external access while letting Red Hat authorized IPs acccess (via `rh-api`, as proposed in this comment).